diff --git a/project/jni/application/simutrans/AndroidAppSettings.cfg b/project/jni/application/simutrans/AndroidAppSettings.cfg
index 67ccbfaa9..e1d2058c9 100644
--- a/project/jni/application/simutrans/AndroidAppSettings.cfg
+++ b/project/jni/application/simutrans/AndroidAppSettings.cfg
@@ -290,7 +290,7 @@ MultiABI='armeabi-v7a arm64-v8a x86 x86_64'
 # Optional shared libraries to compile - removing some of them will save space
 # MP3 patents are expired, but libmad license is GPL, not LGPL
 # Available libraries: mad (GPL-ed!) sdl_mixer sdl_image sdl_ttf sdl_net sdl_blitpool sdl_gfx sdl_sound intl xml2 lua jpeg png ogg flac tremor vorbis freetype xerces curl theora fluidsynth lzma lzo2 mikmod openal timidity zzip bzip2 yaml-cpp python boost_date_time boost_filesystem boost_iostreams boost_program_options boost_regex boost_signals boost_system boost_thread glu avcodec avdevice avfilter avformat avresample avutil swscale swresample bzip2 
-CompiledLibraries="jpeg png freetype bzip2 ssl crypto curl zip fluidsynth c++_shared"
+CompiledLibraries="png freetype bzip2 ssl crypto curl zip fluidsynth c++_shared"
 
 # Application uses custom build script AndroidBuild.sh instead of Android.mk (y) or (n)
 CustomBuildScript=y
diff --git a/project/jni/application/simutrans/AndroidBuild.sh b/project/jni/application/simutrans/AndroidBuild.sh
index 18ba7761d..331afa45d 100755
--- a/project/jni/application/simutrans/AndroidBuild.sh
+++ b/project/jni/application/simutrans/AndroidBuild.sh
@@ -13,20 +13,14 @@ echo BACKEND=sdl2 >> config.$1.txt
 echo USE_SOFTPOINTER=1 >> config.$1.txt
 echo USE_FREETYPE=1 >> config.$1.txt
 echo USE_FLUIDSYNTH_MIDI=1 >> config.$1.txt
+echo WITH_REVISION=1 >> config.$1.txt
+echo MSG_LEVEL=3 >> config.$1.txt
+echo DEBUG=1 >> config.$1.txt
 
 cmake -E copy_if_different config.$1.txt config.$1
 
-# check first if svn repo; otherwise fallback to git repo
-svn info --show-item revision
-if [ $? -eq 0 ]; then
-    echo "#define REVISION `svn info --show-item revision`" > revision.h.txt
-else
-    echo "#define REVISION `git svn find-rev $(git log origin --max-count 1 --pretty=format:%H)`" > revision.h.txt
-fi
-cmake -E copy_if_different revision.h.txt revision.h
-
-env CFLAGS="-fpermissive" \
-	LDFLAGS="-L`pwd`/../../../../obj/local/$1" \
+env CFLAGS="-fpermissive -fvisibility=hidden -fvisibility-inlines-hidden -ffunction-sections -fdata-sections" \
+	LDFLAGS="-L`pwd`/../../../../obj/local/$1 -Wl,--gc-sections -s " \
 	PATH=`pwd`/../..:$PATH \
 	../../setEnvironment-$1.sh sh -c " \
 		make -j8 CFG=$1 && \
