diff --git a/project/jni/application/simutrans/AndroidAppSettings.cfg b/project/jni/application/simutrans/AndroidAppSettings.cfg
index 3dbf10872..f31dfa8b2 100644
--- a/project/jni/application/simutrans/AndroidAppSettings.cfg
+++ b/project/jni/application/simutrans/AndroidAppSettings.cfg
@@ -7,10 +7,10 @@ AppName="Simutrans"
 AppFullName=com.simutrans
 
 # Application version code (integer)
-AppVersionCode=1210
+AppVersionCode=1220
 
 # Application user-visible version name (string)
-AppVersionName="121.0"
+AppVersionName="122.0"
 
 # Specify path to download application data in zip archive in the form "Description|URL|MirrorURL^Description2|URL2|MirrorURL2^..."
 # If you'll start Description with '!' symbol it will be enabled by default, '!!' will also hide the entry from the menu, so it cannot be disabled
@@ -74,7 +74,7 @@ UseGl4es=
 SwVideoMode=y
 
 # Application video output will be resized to fit into native device screen (y)/(n)
-SdlVideoResize=n
+SdlVideoResize=y
 
 # Application resizing will keep 4:3 aspect ratio, with black bars at sides (y)/(n)
 SdlVideoResizeKeepAspect=n
@@ -314,7 +314,7 @@ AppSubdirsBuild=''
 AppBuildExclude=''
 
 # Application command line parameters, including app name as 0-th param
-AppCmdline='simutrans -use_workdir -screensize 800x600 -fullscreen'
+AppCmdline='simutrans -use_workdir -screensize 800x480 -fullscreen'
 
 # Screen size is used by Google Play to prevent an app to be installed on devices with smaller screens
 # Minimum screen size that application supports: (s)mall / (m)edium / (l)arge
